mmcv/ops/sync_bn.py

```diff
@@ -30,41 +30,78 @@ def symbolic(g, input, running_mean, running_var, weight, bias, momentum,
+        self.stats_mode = stats_mode
 
         assert isinstance(
                    input, (torch.HalfTensor, torch.FloatTensor,
                            torch.cuda.HalfTensor, torch.cuda.FloatTensor)), \
                f'only support Half or Float Tensor, but {input.type()}'
-        output = torch.empty_like(input)
-        input3d = input.view(input.size(0), input.size(1), -1)
+        output = torch.zeros_like(input)
+        input3d = input.flatten(start_dim=2)
         output3d = output.view_as(input3d)
+        num_channels = input3d.size(1)
 
-        mean = torch.empty(
-            input3d.size(1), dtype=torch.float, device=input3d.device)
-        var = torch.empty(
-            input3d.size(1), dtype=torch.float, device=input3d.device)
-        norm = torch.empty_like(
+        # make sure mean/var/norm/std are initialized as zeros
+        # torch.empty does not guarantee that
+        mean = torch.zeros(
+            num_channels, dtype=torch.float, device=input3d.device)
+        var = torch.zeros(
+            num_channels, dtype=torch.float, device=input3d.device)
+        norm = torch.zeros_like(
             input3d, dtype=torch.float, device=input3d.device)
-        std = torch.empty(
-            input3d.size(1), dtype=torch.float, device=input3d.device)
+        std = torch.zeros(
+            num_channels, dtype=torch.float, device=input3d.device)
+
+        batch_size = input3d.size(0)
+        if batch_size > 0:
+            ext_module.sync_bn_forward_mean(input3d, mean)
+            batch_flag = torch.ones([1], device=mean.device, dtype=mean.dtype)
+        else:
+            # skip updating mean and leave it as zeros when the input is empty
+            batch_flag = torch.zeros([1], device=mean.device, dtype=mean.dtype)
+
+        # sync mean and the batch flag
+        vec = torch.cat([mean, batch_flag])
+        if self.stats_mode == 'N':
+            vec *= batch_size
 
-        ext_module.sync_bn_forward_mean(input3d, mean)
         if self.group_size > 1:
-            dist.all_reduce(mean, group=self.group)
-            mean /= self.group_size
-        ext_module.sync_bn_forward_var(input3d, mean, var)
+            dist.all_reduce(vec, group=self.group)
+        # need total batch to decide whether to update the momentum
+        total_batch = vec[-1].detach()
+        mean = vec[:num_channels]
+
+        if self.stats_mode == '':
+            mean = mean / self.group_size
+        else:
+            mean = mean / total_batch.clamp(min=1)
+
+        if batch_size > 0:
+            ext_module.sync_bn_forward_var(input3d, mean, var)
+        # skip updating var and leave it as zeros when the input is empty
+
+        if self.stats_mode == 'N':
+            var *= batch_size
         if self.group_size > 1:
             dist.all_reduce(var, group=self.group)
+
+        if self.stats_mode == '':
             var /= self.group_size
+        else:
+            var /= total_batch.clamp(min=1)
```

Comment:
Use `elif self.stats_mode == 'N'` would be better, since we might add other stats_mode in the future?
 A `NotImplementError` could be placed in `else` block.